exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -404,21 +381,23 @@ public static Builder builder() {
 
     private EmulatedNodeType nodeType = EmulatedNodeType.VALIDATOR;
     private short validatorCount = 1;
-    private final List<Class<? extends ServiceModule>> services = new ArrayList<>();
-    private TimeProvider timeProvider;
+    private Multimap<ServiceArtifactId, ServiceSpec> services = ArrayListMultimap.create();
+    private HashMap<ServiceArtifactId, String> serviceArtifactFilenames = new HashMap<>();
+    private TimeServiceSpec timeServiceSpec;
 
-    private Builder() {
-    }
+    private Builder() {}
 
     /**
      * Returns a copy of this TestKit builder.
      */
     public Builder copy() {
-      return new Builder()
+      Builder builder = new Builder()
           .withNodeType(nodeType)
-          .withServices(services)
-          .withValidators(validatorCount)
-          .withTimeService(timeProvider);
+          .withValidators(validatorCount);
+      builder.timeServiceSpec = timeServiceSpec;
+      builder.services = services;
```

Comment:
Is an implementation making a shallow copy of a builder acceptable?

Reply:
As of now - yes. It is only used in TestKitExtension (I think we can make it package-private?) and it did a shallow copy before these changes as well. Should we keep it public and make a deep copy?

Reply:
I don't think we shall make it public. I'd document that we intentionally perform a shallow copy and it is OK.